library/core/src/any.rs

```diff
@@ -689,7 +690,38 @@ impl TypeId {
     #[stable(feature = "rust1", since = "1.0.0")]
     #[rustc_const_unstable(feature = "const_type_id", issue = "77125")]
     pub const fn of<T: ?Sized + 'static>() -> TypeId {
-        TypeId { t: intrinsics::type_id::<T>() }
+        #[cfg(bootstrap)]
+        let t = intrinsics::type_id::<T>() as u128;
+        #[cfg(not(bootstrap))]
+        let t: u128 = intrinsics::type_id::<T>();
+        TypeId { t }
+    }
+}
+
+#[stable(feature = "rust1", since = "1.0.0")]
+impl hash::Hash for TypeId {
+    #[inline]
+    fn hash<H: ~const hash::Hasher>(&self, state: &mut H) {
+        // Hash the lower 64 bits as a `u64`. This avoids breaking some
+        // libraries in the ecosystem, and arguably is the right call anyway,
+        // because:
+        //
+        // - It's correct for a `Hash` implementation to only hash a subset of
+        //   `self`; doing so is consistent with an `Eq` which considers the
+        //   whole value (as ours does).
+        //
+        // - The hashing algorithm which backs `TypeId` is expected to be
+        //   unbiased and high quality, meaning further mixing would be somewhat
+        //   redundant compared to choosing 64 bits arbitrarially (like we do).
+        //
+        // - `Hasher::finish` returns a u64 anyway, so the extra entropy we'd
+        //   get from hashing the full value would probably not be useful
+        //   (especially given the previous point about the lower 64 bits being
+        //   high quality on their own).
+        //
+        // That said, this is a bit of a hack. (FIXME(thom): nominate this for
```

Comment:
Don't land without me doing this (but it probably would be premature before the MCP completes).

Reply:
ping: the MCP was completed, do you still want to nominate this for `T-libs`?

(IMO this impl can be changed after/asynchronously to this PR — as discussed here previously, this implementation is totally fine for the `Hash` trait)